File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int count = 0;
    public int goodNodes(TreeNode root) {

        
        helper(root, Integer.MIN_VALUE);
        return count;
    }
    
    public void helper(TreeNode root, int prev){
        if(root == null)
            return;
        
        if(prev <= root.val)
            ++count;
        
        int max = Math.max(prev, root.val);
        helper(root.left, max);
        helper(root.right, max);
    }
}

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int: 
        '''
            Use dfs 
            If a node is greater than the previous maxvalue, set the count to 1
            Update the previous max value if the node is greater than it
            Add the dfs of the left and right children to the count
        '''
        return self.dfs(root, float('-inf'))
       
    
    def dfs(self, node, prevMax):
        if not node:
            return 0
        
        res = 1 if node.val >= prevMax else 0
        
        prevMax = max(prevMax, node.val)
        res += self.dfs(node.left, prevMax)
        res += self.dfs(node.right, prevMax)
        return res
        
        